security/pkg/nodeagent/caclient/providers/keyfactor/client.go

```diff
@@ -0,0 +1,278 @@
+	}
+
+	keyFactorCAClientLog.Infof("Create keyfactor metadatas: %v", keyfactorConfig.Metadata)
+
+	customMetadatas := make(map[string]string)
+
+	for fieldKey, aliasName := range keyfactorConfig.Metadata {
+		if aliasName == "" {
+			keyFactorCAClientLog.Infof("Skip metadata fieldName: %v", fieldKey)
+			continue
+		}
+		switch fieldKey {
+		case "Cluster":
+			customMetadatas[aliasName] = metadata.ClusterID
+		case "Service":
+			serviceName := metadata.PodName
+			if splitPodName := strings.Split(metadata.PodName, "-"); len(splitPodName) > 2 {
+				// example: service-name-A-v1-roiwe0239-24jfef9 => service-name-A-v1
+				arrayOfServiceNames := splitPodName[0 : len(splitPodName)-2]
+				serviceName = strings.Join(arrayOfServiceNames, "-")
+			}
+			customMetadatas[aliasName] = serviceName
+		case "PodName":
+			customMetadatas[aliasName] = metadata.PodName
+		case "PodNamespace":
+			customMetadatas[aliasName] = metadata.PodNamespace
+		case "PodIP":
+			customMetadatas[aliasName] = metadata.PodIP
+		case "TrustDomain":
+			customMetadatas[aliasName] = metadata.TrustDomain
+		}
+	}
+
+	c := &KeyFactorCAClient{
+		CaEndpoint:      endpoint,
+		EnableTLS:       enableTLS,
+		CustomMetadatas: customMetadatas,
+		Metadata:        metadata,
+		ClientOptions:   keyfactorConfig,
+	}
+
+	if !enableTLS {
+
+		transport := &http.Transport{
+			TLSClientConfig: &tls.Config{
+				InsecureSkipVerify: true,
+			},
+		}
+		c.Client = &http.Client{
+			Transport: transport,
+			Timeout:   30 * time.Second,
+		}
+	} else {
+
+		// Load the system default root certificates.
+		pool, err := x509.SystemCertPool()
+		if err != nil {
+			keyFactorCAClientLog.Errorf("could not get SystemCertPool: %v", err)
+			return nil, fmt.Errorf("could not get SystemCertPool: %v", err)
+		}
+
+		if pool == nil {
+			log.Info("System cert pool is nil, create a new cert pool")
+			pool = x509.NewCertPool()
+		}
+
+		if len(rootCert) > 0 {
+			ok := pool.AppendCertsFromPEM(rootCert)
+			if !ok {
+				return nil, fmt.Errorf("invalid root-cert.pem: %v", string(rootCert))
+			}
+		}
+
+		tlsConfig := &tls.Config{
+			RootCAs: pool,
+		}
+
+		transport := &http.Transport{TLSClientConfig: tlsConfig}
+
+		c.Client = &http.Client{
+			Transport: transport,
+			Timeout:   30 * time.Second,
+		}
+	}
+
+	return c, nil
+}
+
+// CSRSign calls KeyFactor CA to sign a CSR.
+func (cl *KeyFactorCAClient) CSRSign(ctx context.Context, reqID string, csrPEM []byte, subjectID string,
+	certValidTTLInSec int64) ([]string /*PEM-encoded certificate chain*/, error) {
+
+	bytesRepresentation, err := json.Marshal(keyfactorRequestPayload{
+		CSR:                  string(csrPEM),
+		CertificateAuthority: cl.ClientOptions.CaName,
+		IncludeChain:         true,
+		Template:             cl.ClientOptions.CaTemplate,
+		TimeStamp:            time.Now().Format(time.RFC3339),
+		Metadata:             cl.CustomMetadatas,
+		SANs: san{
+			DNS: []string{cl.Metadata.ClusterID},
+			IP4: []string{cl.Metadata.PodIP},
```

Comment:
I don't think Citadel signer does this. Why is this needed? IP based security does not seem like it belongs here

Reply:
It's possible the design of Istio makes this unnecessary in a typical deployment. However, any client attempting to establish a TLS connection to the pod by IP address should be looking at this SAN to verify identity, per [RFC 2818](https://tools.ietf.org/html/rfc2818#section-3.1). Including the IP address hosting the certificate may also be useful for associating network traces and logs for auditing or other purposes. Note also that even if the IP address SAN is included along with a CSR in the request, the Microsoft ADCS CA that most of our customers use behind our platform can be configured not to include this in the signed cert, based on the ADCS certificate template used.

Reply:
The competing concern would be that exposing pod IP info to other resources through this channel would leak internal configuration that may present a security risk (e.g. to a non-Istio user that is able to read the CA database).

Reply:
I am far from an expert here. @myidpt would love your input here. Thanks for the references btw